### PR TITLE
new GameManager and GameStateManager

### PR DESCRIPTION
**Pull request by Hanson-Sun**
added a new GameManager class

since its unity, the game manager is a singleton unity object that extends MonoBehavior 

GameStateManager is not fully implemented as it has no use currently

### CODE REVIEW

File: Assets/Scripts/SceneUtility/GameManager.cs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+{
+    private static GameManager instance = null;
+    private readonly List<Player> players;
+    private readonly List<Projectile> projectiles;
+    private readonly List<Trap> traps;
+    private double time;
+
+    public static GameManager Instance
+    {
+        get
```
Comment (leo-cf-tian) [inline Assets/Scripts/SceneUtility/GameManager.cs:16]: consider using a lambda function lol

File: Assets/Scripts/SceneUtility/GameManager.cs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        {
+            if (instance == null)
+            {
+                GameObject gameManager = new GameObject();
```
Comment (leo-cf-tian) [inline Assets/Scripts/SceneUtility/GameManager.cs:20]: consider using null coalescing operators

File: Assets/Scripts/SceneUtility/GameManager.cs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    {
+        this.players = new List<Player>();
+        this.projectiles = new List<Projectile>();
+        this.traps = new List<Trap>();
+        this.time = 0;
+    }
+
+    private void Awake()
+    {
+        if (instance != null && instance != this)
```
Comment (leo-cf-tian) [inline Assets/Scripts/SceneUtility/GameManager.cs:37]: not something you need to change, but you could do without parentheses probably

File: Assets/Scripts/SceneUtility/GameStateManager.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    END
+}
+
+public class GameStateManager
+{
+    private static GameStateManager instance = null;
+
+    public static GameStateManager Instance
+    {
+        get
```
Comment (leo-cf-tian) [inline Assets/Scripts/SceneUtility/GameStateManager.cs:21]: same comment about lambda functions

File: Assets/Scripts/SceneUtility/GameStateManager.cs [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        {
+            instance ??= new GameStateManager();
```
Comment (leo-cf-tian) [inline Assets/Scripts/SceneUtility/GameStateManager.cs:23]: you use it here lol

**Review by leo-cf-tian — CHANGES_REQUESTED** [review]
Pretty good, could use a few simplifications though.

File: Assets/Scripts/SceneUtility/GameManager.cs (hunk 2 above)
Comment (Hanson-Sun) [reply Assets/Scripts/SceneUtility/GameManager.cs:20]: he wants the ??

Comment (leo-cf-tian) [reply Assets/Scripts/SceneUtility/GameManager.cs:20]: yep, you could probably use it with the lambda function too to make it way shorter

File: Assets/Scripts/SceneUtility/GameStateManager.cs (hunk 5 above)
Comment (Hanson-Sun) [reply Assets/Scripts/SceneUtility/GameStateManager.cs:23]: * the linter did it

Comment (leo-cf-tian) [reply Assets/Scripts/SceneUtility/GameStateManager.cs:23]: You should do it to!

**Review by anjx123 — APPROVED** [review]
Manager looks good tbh, its probably good enough for now since we have no use for so much functionality.

**Review by leo-cf-tian — APPROVED** [review]

**Pull request merged**